src/Microsoft.Extensions.HealthChecks/HealthCheckBuilder.cs

```diff
@@ -8,24 +8,109 @@ namespace Microsoft.Extensions.HealthChecks
         {
-            _checks = new Dictionary<string, IHealthCheck>(StringComparer.OrdinalIgnoreCase);
+            _checksByName = new Dictionary<string, CachedHealthCheck>(StringComparer.OrdinalIgnoreCase);
+            _currentGroup = new HealthCheckGroup(string.Empty, CheckStatus.Unhealthy);
+            _groups = new Dictionary<string, HealthCheckGroup>(StringComparer.OrdinalIgnoreCase)
+            {
+                [string.Empty] = _currentGroup
+            };
+
             DefaultCacheDuration = TimeSpan.FromMinutes(5);
         }
 
-        public IReadOnlyDictionary<string, IHealthCheck> Checks => _checks;
+        /// <summary>
+        /// This constructor should only be used when creating a grouped health check builder.
+        /// </summary>
+        public HealthCheckBuilder(HealthCheckBuilder rootBuilder, HealthCheckGroup currentGroup)
+        {
+            Guard.ArgumentNotNull(nameof(rootBuilder), rootBuilder);
+            Guard.ArgumentNotNull(nameof(currentGroup), currentGroup);
+
+            _checksByName = rootBuilder._checksByName;
+            _currentGroup = currentGroup;
+            _groups = rootBuilder._groups;
+
+            DefaultCacheDuration = rootBuilder.DefaultCacheDuration;
+        }
+
+        /// <summary>
+        /// Gets the registered checks, indexed by check name.
+        /// </summary>
+        public IReadOnlyDictionary<string, CachedHealthCheck> ChecksByName => _checksByName;
 
+        /// <summary>
+        /// Gets the current default cache duration used when registering checks.
+        /// </summary>
         public TimeSpan DefaultCacheDuration { get; private set; }
 
-        public HealthCheckBuilder AddCheck(string name, IHealthCheck check)
+        /// <summary>
+        /// Gets the registered groups, indexed by group name. The root group's name is <see cref="string.Empty"/>.
+        /// </summary>
+        public IReadOnlyDictionary<string, HealthCheckGroup> Groups => _groups;
+
+        /// <summary>
+        /// Registers a health check type that will later be resolved via dependency
+        /// injection.
+        /// </summary>
+        public HealthCheckBuilder AddCheck<TCheck>(string checkName, TimeSpan cacheDuration) where TCheck : class, IHealthCheck
         {
-            Guard.ArgumentNotNullOrWhitespace(nameof(name), name);
+            Guard.ArgumentNotNullOrWhitespace(nameof(checkName), checkName);
+
+            var namedCheck = CachedHealthCheck.FromType(checkName, cacheDuration, typeof(TCheck));
+
+            _checksByName.Add(checkName, namedCheck);
+            _currentGroup.ChecksInternal.Add(namedCheck);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Registers a concrete health check to the builder.
+        /// </summary>
+        public HealthCheckBuilder AddCheck(string checkName, TimeSpan cacheDuration, IHealthCheck check)
+        {
+            Guard.ArgumentNotNullOrWhitespace(nameof(checkName), checkName);
```

Comment:
Unless, that is, there's some actual technical reason why it's a bad idea. But presumably this string is always represented as a structural string, so it ought to transport with full fidelity no matter where it goes.